core/opregion.c

```diff
@@ -38,571 +41,327 @@ void lai_write_opregion(lai_nsnode_t *field, lai_variable_t *source) {
     lai_panic("undefined field write: %s", lai_stringify_node_path(field));
 }
 
-void lai_read_field(lai_variable_t *destination, lai_nsnode_t *field) {
-    struct lai_instance *instance = lai_current_instance();
-    lai_nsnode_t *opregion = field->fld_region_node;
+// this assumes little endian
+static void lai_buffer_put_at(uint8_t *buffer, uint64_t value, size_t bit_offset, size_t num_bits) {
+    size_t in_byte_offset = bit_offset & 7;
+    size_t byte_offset = bit_offset >> 3;
+    size_t progress = 0;
+    while (progress < num_bits) {
+        uint8_t byte = (value >> progress) & 0xFF;
 
-    uint64_t offset, value, mask;
-    size_t bit_offset;
+        uint8_t low_byte = byte << in_byte_offset;
+        uint8_t hi_byte = byte >> (8 - in_byte_offset);
 
-    mask = ((uint64_t)1 << field->fld_size);
-    mask--;
-    offset = field->fld_offset / 8;
-    void *mmio;
+        buffer[byte_offset + progress / 8] |= low_byte;
+        buffer[byte_offset + 1 + progress / 8] |= hi_byte;
```

Comment:
This writes past the end of the buffer. Consider a 1-byte buffer where you want to write 1 bit.

Just rewrite it to do the usual bit-progress loop: write `min(num_bits - progress, 8 - in_byte_offset)` bits at a time, where `in_byte_offset` is recalculated as `(bit_offset + progress) & 7` in each iteration.